File: schemas/content-moderation/drizzle/auto_detection_results.ts
// auto_detection_results: Automated content analysis results.
// See README.md for full design rationale.

import {
  pgEnum,
  pgTable,
  uuid,
  text,
  numeric,
  boolean,
  jsonb,
  timestamp,
  index,
} from "drizzle-orm/pg-core";
import { sql } from "drizzle-orm";
import { moderationQueueItems } from "./moderation_queue_items";
import { moderationRules } from "./moderation_rules";

export const detectionContentTypeEnum = pgEnum("detection_content_type", [
  "post",
  "comment",
  "message",
  "user",
  "media",
]);

export const detectionMethodEnum = pgEnum("detection_method", [
  "ml_classifier",
  "hash_match",
  "keyword",
  "regex",
  "blocklist",
]);

export const autoDetectionResults = pgTable(
  "auto_detection_results",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    contentType: detectionContentTypeEnum("content_type").notNull(), // What type of content was analyzed.
    contentId: text("content_id").notNull(), // What was analyzed.
    queueItemId: uuid("queue_item_id").references(
      () => moderationQueueItems.id,
      { onDelete: "set null" }
    ), // Queue item created/updated by this detection.
    detectionMethod: detectionMethodEnum("detection_method").notNull(), // Type of detection that produced this result.
    detectionSource: text("detection_source").notNull(), // Specific detector name (e.g., "perspective", "openai", "photodna").
    category: text("category"), // Detected violation category (e.g., "toxicity", "hate_speech", "csam_hash_match").
    confidenceScore: numeric("confidence_score"), // Detection confidence, 0.00 to 1.00. Nullable: some methods (keyword match) are binary.
    matchedValue: text("matched_value"), // What triggered the match (keyword, regex pattern, hash ID, etc.).
    isActionable: boolean("is_actionable").notNull().default(false), // Whether this result met the threshold for automated action.
    metadata: jsonb("metadata").default(sql`'{}'`), // Detection-specific extra data (per-category scores, hash distances, etc.).
    ruleId: uuid("rule_id").references(() => moderationRules.id, {
      onDelete: "set null",
    }), // The rule that triggered this detection, if applicable.

    // Immutable. Detections are append-only — no updated_at.
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("idx_auto_detection_results_content").on(
      table.contentType,
      table.contentId
    ),
    index("idx_auto_detection_results_queue_item_id").on(table.queueItemId),
    index("idx_auto_detection_results_detection_method").on(
      table.detectionMethod
    ),
    index("idx_auto_detection_results_detection_source").on(
      table.detectionSource
    ),
    index("idx_auto_detection_results_is_actionable").on(table.isActionable),
    index("idx_auto_detection_results_created_at").on(table.createdAt),
  ]
);

File: schemas/content-moderation/spacetimedb/auto_detection_results.rs
// auto_detection_results: Automated content analysis results.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

/// Type of content that was analyzed.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum AutoDetectionContentType {
    Post,
    Comment,
    Message,
    User,
    Media,
}

/// Type of detection that produced this result.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum AutoDetectionMethod {
    MlClassifier,
    HashMatch,
    Keyword,
    Regex,
    Blocklist,
}

#[spacetimedb::table(name = auto_detection_results, public)]
pub struct AutoDetectionResult {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub content_type: AutoDetectionContentType,

    #[index(btree)]
    pub content_id: String, // What was analyzed.

    #[index(btree)]
    pub queue_item_id: Option<String>, // UUID — FK → moderation_queue_items.id (set null)

    #[index(btree)]
    pub detection_method: AutoDetectionMethod,

    #[index(btree)]
    pub detection_source: String, // Specific detector name (e.g., "perspective", "photodna").
                                  // String, not enum — new detectors added without schema changes.

    pub category: Option<String>, // Detected violation category (e.g., "toxicity", "hate_speech").
                                  // String to support detector-specific category names.

    pub confidence_score: Option<f64>, // Detection confidence, 0.00 to 1.00.
                                       // Nullable: some methods (keyword match) are binary, not scored.

    pub matched_value: Option<String>, // What triggered the match (keyword, regex pattern, hash ID).

    #[index(btree)]
    pub is_actionable: bool, // Whether this result met the threshold for automated action.

    pub metadata: Option<String>, // JSON stored as string

    pub rule_id: Option<String>, // UUID — FK → moderation_rules.id (set null)

    #[index(btree)]
    pub created_at: Timestamp, // Detections are immutable. No updated_at.
}
